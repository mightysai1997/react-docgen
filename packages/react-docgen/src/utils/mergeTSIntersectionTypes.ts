import type {
  TSFunctionSignatureType,
  TypeDescriptor,
} from '../Documentation.js';

/**
 * Merges two TSFunctionSignatureType types into one.
 *
 * @example
 * const existingType = {
 *   "key": "children",
 *   "value": {
 *     "name": "ReactNode",
 *     "required": true,
 *   },
 * };
 * const newType = {
 *   "key": "children",
 *   "value": {
 *     "name": "never",
 *     "required": false,
 *   },
 * };
 *
 * return {
 *   "key": "children",
 *   "value": {
 *     "name": "ReactNode",
 *     "required": false,
 *   },
 * };
 */
export default (
  newType: TypeDescriptor<TSFunctionSignatureType>,
  existingType: TypeDescriptor<TSFunctionSignatureType>,
): TypeDescriptor<TSFunctionSignatureType> => {
  const newTypeIsForbiddenToUse = ['never'].includes(newType.name);

  if (newTypeIsForbiddenToUse) {
    let mergedType = {
      ...existingType,
      required:
        newType.required === false || existingType.required === false
          ? false
          : existingType.required,
    };

    if ('nullable' in existingType) {
      mergedType = {
        ...mergedType,
        nullable: newType.nullable === true ? true : existingType.nullable,
      };
    }

    return mergedType;
  }

  return {
    ...newType,
    required:
      newType.required === false || existingType.required === false
        ? false
        : existingType.required,
  };
};
