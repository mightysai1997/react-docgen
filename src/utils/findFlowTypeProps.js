/*
 * Copyright (c) 2016, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import type Documentation from '../Documentation';
type SetPropCallback = (documentation: Documentation, path: NodePath) => void

/**
 * Tries to find all props from the flow annotation and calls the setPropCallback for each one.
 * The callback can be called multiple times for the same prop (Union, Intersection).
 */
export default function findFlowTypeProps(documentation: Documentation, path: NodePath, setPropCallback: SetPropCallback): void {
  if (path.node.properties) {
    path.get('properties').each(
      propertyPath => setPropCallback(documentation, propertyPath)
    );
  } else if (path.node.types) {
    path.get('types').each(
      typesPath => findFlowTypeProps(documentation, typesPath, setPropCallback)
    );
  }
}
