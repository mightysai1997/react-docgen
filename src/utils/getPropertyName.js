/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import recast from 'recast';
import getNameOrValue from './getNameOrValue';

const {
  types: { namedTypes: types },
} = recast;

export const COMPUTED_PREFIX = '@computed#';

/**
 * In an ObjectExpression, the name of a property can either be an identifier
 * or a literal (or dynamic, but we don't support those). This function simply
 * returns the value of the literal or name of the identifier.
 */
export default function getPropertyName(propertyPath: NodePath): ?string {
  if (types.ObjectTypeSpreadProperty.check(propertyPath.node)) {
    return getNameOrValue(propertyPath.get('argument').get('id'), false);
  } else if (propertyPath.node.computed) {
    if (types.Identifier.check(propertyPath.node.key)) {
      return `${COMPUTED_PREFIX}${getNameOrValue(
        propertyPath.get('key'),
        false,
      )}`;
    } else if (
      types.Literal.check(propertyPath.node.key) &&
      (typeof propertyPath.node.key.value === 'string' ||
        typeof propertyPath.node.key.value === 'number')
    ) {
      return `${COMPUTED_PREFIX}${propertyPath.node.key.value}`;
    }

    return null;
  }

  return getNameOrValue(propertyPath.get('key'), false);
}
