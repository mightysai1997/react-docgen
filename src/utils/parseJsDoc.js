/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

import doctrine from 'doctrine';

type JsDoc = {
  description: ?string;
  params: Array<{
    name: string;
    description: ?string;
    type: ?{name: string; value?: Array<string>;};
    optional?: boolean;
    rest?:boolean;
  }>;
  returns: ?{
    description: ?string;
    type: ?{name: string; value?: Array<string>;};
  };
};

function getType(tag) {
  var res = null;
  if (!tag.type) {
    return null;
  } else if (typeof tag.type === 'object'){
    res = getType(tag.type);
  } else if (tag.type === 'UnionType') {
    // union type
    res = {name: 'union', value: tag.elements.map(function (element) {
      return element.name;
    })};
  } else if (tag.type === 'AllLiteral') {
    // return {*}
    res =  {name: 'mixed'};
  }
  res = res || {name: tag.name || tag.expression.name};
  if (tag.expression) {
    Object.assign(res, getType(tag.expression));
  }
  return res;
}

function getOptional(tag) {
  if (tag.type && tag.type.type && tag.type.type === 'OptionalType') {
    return true;
  }
  return;
}

function getRest(tag) {
    return (tag.type && tag.type.expression && tag.type.expression.type && tag.type.expression.type === 'RestType') || undefined;
}

// Add jsdoc @return description.
function getReturnsJsDoc(jsDoc) {
  const returnTag = jsDoc.tags.find(
    tag => tag.title === 'return' || tag.title === 'returns'
  );
  if (returnTag) {
    return {
      description: returnTag.description,
      type: getType(returnTag),
    };
  }
  return null;
}

// Add jsdoc @param descriptions.
function getParamsJsDoc(jsDoc) {
  if (!jsDoc.tags) {
    return [];
  }
  return jsDoc.tags
    .filter(tag => tag.title === 'param')
    .map(tag => {
      return {
        name: tag.name,
        description: tag.description,
        type: getType(tag),
        optional: getOptional(tag),
        rest: getRest(tag),
      };
    });
}

export default function parseJsDoc(docblock: string): JsDoc {
  const jsDoc = doctrine.parse(docblock,{sloppy:true});
  return {
    description: jsDoc.description || null,
    params: getParamsJsDoc(jsDoc),
    returns: getReturnsJsDoc(jsDoc),
  };
}
