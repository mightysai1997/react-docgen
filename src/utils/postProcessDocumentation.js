/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

function postProcessProps(props) {
  // props with default values should not be required
  Object.keys(props).forEach(prop => {
    const propInfo = props[prop];

    if (propInfo.defaultValue) {
      propInfo.required = false;

      // Check and remove any flow type cast.  Use last `:` in case description has `:` in it.
      propInfo.defaultValue.value = propInfo.defaultValue.value.includes(':') ?
        propInfo.defaultValue.value.slice(0, propInfo.defaultValue.value.lastIndexOf(':')) :
        propInfo.defaultValue.value
    }
  });
}

export default function (documentation: any) {
  const props = documentation.props;
  if (props) {
    postProcessProps(props);
  }

  return documentation;
}
