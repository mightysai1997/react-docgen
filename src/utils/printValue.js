/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import strip from 'strip-indent';

function deindent(code: string): string {
  const firstNewLine = code.indexOf('\n');

  return (
    code.slice(0, firstNewLine + 1) +
    // remove indentation from all lines except first.
    strip(code.slice(firstNewLine + 1))
  );
}

function getSrcFromAst(path: NodePath): string {
  do {
    if (path.node.type === 'File') {
      return path.node.__src;
    }
    path = path.parentPath;
  } while (path != null);

  throw new Error('Could not find source attached to File node');
}

/**
 * Prints the given path without leading or trailing comments.
 */
export default function printValue(path: NodePath): string {
  if (path.node.start == null) {
    if (path.node.type === 'Literal') {
      return `"${path.node.value}"`;
    }
    throw new Error(
      `Cannot print raw value for type '${
        path.node.type
      }'. Please report this with an example at https://github.com/reactjs/react-docgen/issues`,
    );
  }
  const src = getSrcFromAst(path);

  return deindent(src.slice(path.node.start, path.node.end));
}
