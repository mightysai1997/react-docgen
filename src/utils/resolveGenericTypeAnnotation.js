/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import isUnreachableFlowType from '../utils/isUnreachableFlowType';
import recast from 'recast';
import resolveToValue from '../utils/resolveToValue';
import { unwrapUtilityType } from './flowUtilityTypes';

const {
  types: { namedTypes: types },
} = recast;

/**
 * Given an React component (stateless or class) tries to find the
 * flow type for the props. If not found or not one of the supported
 * component types returns null.
 */
export default function resolveGenericTypeAnnotation(
  path: NodePath,
): ?NodePath {
  // If the node doesn't have types or properties, try to get the type.
  let typePath = unwrapUtilityType(path);

  if (types.GenericTypeAnnotation.check(typePath.node)) {
    typePath = resolveToValue(typePath.get('id'));
    if (isUnreachableFlowType(typePath)) {
      return;
    }

    typePath = unwrapUtilityType(typePath.get('right'));
  }

  return typePath;
}
