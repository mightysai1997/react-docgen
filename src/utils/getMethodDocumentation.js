/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

import getFlowType from './getFlowType';
import getMethodJSDoc from './getMethodJSDoc';
import getPropertyName from './getPropertyName';
import getTypeAnnotation from './getTypeAnnotation';

type MethodParameter = {
  name: string;
  description?: ?string;
  type?: ?FlowTypeDescriptor;
};
type MethodReturn = {
  description: ?string;
  type: ?FlowTypeDescriptor;
};

type MethodDocumentation = {
  name: string;
  description: ?string;
  modifiers: Array<string>;
  visibility: string;
  params: Array<MethodParameter>;
  return: ?MethodReturn;
};

function getMethodParamsDoc(methodPath, jsDoc) {
  const params = [];
  const functionExpression = methodPath.get('value');

  // Extract param flow types.
  functionExpression.get('params').each(paramPath => {
    let type = null;
    const typePath = getTypeAnnotation(paramPath);
    if (typePath) {
      type = getFlowType(typePath);
    }

    const param = {
      name: paramPath.node.name,
      description: null,
      type,
    };

    params.push(param);
  });

  // Add jsdoc @param descriptions.
  if (jsDoc) {
    jsDoc.tags
      .filter(tag => tag.title === 'param')
      .forEach(tag => {
        const param = params.find(p => p.name === tag.name);
        if (param) {
          param.description = tag.description;
        }
      });
  }

  return params;
}

function getMethodReturnDoc(methodPath, jsDoc) {
  const functionExpression = methodPath.get('value');

  let type = null;
  let description = null;

  // Extract flow return type.
  if (functionExpression.node.returnType) {
    const returnType = getTypeAnnotation(functionExpression.get('returnType'));
    if (returnType) {
      type = getFlowType(returnType);
    }
  }

  // Add jsdoc @return description.
  if (jsDoc) {
    const returnTag = jsDoc.tags.find(tag => tag.title === 'returns');
    if (returnTag) {
      description = returnTag.description;
    }
  }

  if (type || description) {
    return {
      type,
      description,
    };
  }
  return null;
}

function getMethodVisibility(jsDoc) {
  if (jsDoc) {
    for (const tag of jsDoc.tags) {
      switch (tag.title) {
        case 'private':
        case 'public':
        case 'protected':
          return tag.title;
        case 'access':
          return tag.access;
        default:
          break;
      }
    }
  }

  return 'public';
}

function getMethodModifiers(methodPath) {
  const modifiers = [];

  if (methodPath.node.static) {
    modifiers.push('static');
  }

  const functionExpression = methodPath.get('value').node;
  if (functionExpression.generator) {
    modifiers.push('generator');
  }
  if (functionExpression.async) {
    modifiers.push('async');
  }

  return modifiers;
}

export default function getMethodDocumentation(methodPath: NodePath): MethodDocumentation {
  const name = getPropertyName(methodPath);
  const jsDoc = getMethodJSDoc(methodPath);

  return {
    name,
    description: jsDoc && jsDoc.description || null,
    modifiers: getMethodModifiers(methodPath),
    visibility: getMethodVisibility(jsDoc),
    params: getMethodParamsDoc(methodPath, jsDoc),
    return: getMethodReturnDoc(methodPath, jsDoc),
  };
}
