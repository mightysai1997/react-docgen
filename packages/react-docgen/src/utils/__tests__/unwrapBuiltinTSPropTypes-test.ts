import type { TSTypeReference, VariableDeclaration } from '@babel/types';
import { parseTypescript } from '../../../tests/utils';
import unwrapBuiltinTSPropTypes from '../unwrapBuiltinTSPropTypes.js';
import { describe, expect, test } from 'vitest';
import type { NodePath } from '@babel/traverse';

describe('unwrapBuiltinTSPropTypes', () => {
  test('React.PropsWithChildren', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import React from 'react';
         var foo: React.PropsWithChildren<Props>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('React.PropsWithoutRef', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import React from 'react';
         var foo: React.PropsWithoutRef<Props>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('React.PropsWithRef', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import React from 'react';
         var foo: React.PropsWithRef<Props>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('multiple', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import React from 'react';
         var foo: React.PropsWithChildren<React.PropsWithRef<Props>>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('does not follow reassignment', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import React from 'react';
         type bar = React.PropsWithRef<Props>
         var foo: React.PropsWithChildren<bar>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('with require', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `const React = require('react');
         var foo: React.PropsWithRef<Props>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });

  test('with named import', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(
        `import { PropsWithRef } from 'react';
         var foo: PropsWithRef<Props>`,
      )
      .get(
        'declarations.0.id.typeAnnotation.typeAnnotation',
      ) as NodePath<TSTypeReference>;

    expect(unwrapBuiltinTSPropTypes(path)).toMatchSnapshot();
  });
});
