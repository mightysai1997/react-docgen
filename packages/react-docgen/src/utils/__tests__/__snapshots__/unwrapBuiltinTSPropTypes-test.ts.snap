// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unwrapBuiltinTSPropTypes > React.PropsWithChildren 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > React.PropsWithRef 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > React.PropsWithoutRef 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > does not follow reassignment 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "bar",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > multiple 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > with named import 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;

exports[`unwrapBuiltinTSPropTypes > with require 1`] = `
Node {
  "type": "TSTypeReference",
  "typeName": Node {
    "name": "Props",
    "type": "Identifier",
  },
}
`;
